File: packages/protocol/python/person_protocol/framing.py
"""Newline-delimited JSON framing with a hard size ceiling."""

from __future__ import annotations

import json
from typing import Any

MAX_FRAME_BYTES = 1024 * 1024


class FrameError(ValueError):
    """A frame could not be encoded or decoded."""


def encode_frame(message: Any) -> str:
    line = json.dumps(message, separators=(",", ":"), allow_nan=False)
    if "\n" in line:
        raise FrameError("Encoded message contains a newline")
    if len(line.encode("utf-8")) > MAX_FRAME_BYTES:
        raise FrameError("Encoded message exceeds the frame size limit")
    return line + "\n"


def decode_frame(line: str) -> dict[str, Any]:
    if len(line.encode("utf-8")) > MAX_FRAME_BYTES:
        raise FrameError("frame exceeds the size limit")
    try:
        value = json.loads(line)
    except json.JSONDecodeError as error:
        raise FrameError(f"frame is not valid JSON: {error}") from error
    if not isinstance(value, dict):
        raise FrameError("frame is not a JSON object")
    return value


class LineReader:
    """Incremental splitter that refuses to grow past the frame limit."""

    def __init__(self, limit: int = MAX_FRAME_BYTES) -> None:
        self._buffer = ""
        self._limit = limit

    def push(self, chunk: str) -> tuple[list[str], str | None]:
        self._buffer += chunk
        lines: list[str] = []
        while "\n" in self._buffer:
            line, self._buffer = self._buffer.split("\n", 1)
            if line.strip():
                lines.append(line)
        if len(self._buffer.encode("utf-8")) > self._limit:
            self._buffer = ""
            return lines, "unterminated frame exceeded the size limit"
        return lines, None

    @property
    def pending(self) -> str:
        return self._buffer

File: packages/protocol/ts/framing.ts
import { ProtocolError } from "./validator.ts";

/** Hard ceiling for a single framed message. Anything larger is rejected unread. */
export const MAX_FRAME_BYTES = 1024 * 1024;

/**
 * Newline-delimited JSON framing.
 *
 * Framing is deliberately dumb: one JSON object per line, no embedded
 * newlines, one hard size cap. A malformed frame must never be able to stall,
 * desynchronise, or crash the runtime that reads it.
 */
export function encodeFrame(message: unknown): string {
  const line = JSON.stringify(message);
  if (line === undefined)
    throw new ProtocolError("Message is not JSON serialisable");
  if (line.includes("\n"))
    throw new ProtocolError("Encoded message contains a newline");
  if (Buffer.byteLength(line, "utf8") > MAX_FRAME_BYTES)
    throw new ProtocolError("Encoded message exceeds the frame size limit");
  return `${line}\n`;
}

export interface FrameResult {
  ok: boolean;
  value?: unknown;
  error?: string;
}

export function decodeFrame(line: string): FrameResult {
  if (Buffer.byteLength(line, "utf8") > MAX_FRAME_BYTES)
    return { ok: false, error: "frame exceeds the size limit" };
  try {
    const value: unknown = JSON.parse(line);
    if (typeof value !== "object" || value === null || Array.isArray(value))
      return { ok: false, error: "frame is not a JSON object" };
    return { ok: true, value };
  } catch (error) {
    return {
      ok: false,
      error: `frame is not valid JSON: ${(error as Error).message}`,
    };
  }
}

/**
 * Incremental line splitter with a bounded buffer.
 *
 * `push` returns complete lines. When the peer sends an unterminated stream
 * larger than the frame limit the reader reports an error and drops the
 * buffer rather than growing without bound.
 */
export class LineReader {
  #buffer = "";
  readonly #limit: number;

  constructor(limit: number = MAX_FRAME_BYTES) {
    this.#limit = limit;
  }

  push(chunk: string): { lines: string[]; error?: string } {
    this.#buffer += chunk;
    const lines: string[] = [];
    let index = this.#buffer.indexOf("\n");
    while (index >= 0) {
      const line = this.#buffer.slice(0, index);
      this.#buffer = this.#buffer.slice(index + 1);
      if (line.trim().length > 0) lines.push(line);
      index = this.#buffer.indexOf("\n");
    }
    if (Buffer.byteLength(this.#buffer, "utf8") > this.#limit) {
      this.#buffer = "";
      return { lines, error: "unterminated frame exceeded the size limit" };
    }
    return { lines };
  }

  get pending(): string {
    return this.#buffer;
  }
}
